File: java/src/main/java/com/hydrapayments/sdk/HydraClientBuilder.java
package com.hydrapayments.sdk;

import java.net.http.HttpClient;
import java.time.Duration;

public class HydraClientBuilder {

    private String baseUrl = "http://localhost:8080";
    private String apiKey = "";
    private String secretKey = "";
    private String defaultCurrency = "GBP";
    private String locale = "en";
    private Duration timeout = Duration.ofSeconds(30);
    private HttpClient httpClient;

    public HydraClientBuilder baseUrl(String baseUrl) {
        if (baseUrl != null) {
            this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        }
        return this;
    }

    public HydraClientBuilder apiKey(String apiKey) {
        this.apiKey = apiKey;
        return this;
    }

    public HydraClientBuilder secretKey(String secretKey) {
        this.secretKey = secretKey;
        return this;
    }

    public HydraClientBuilder withDefaultCurrency(String defaultCurrency) {
        this.defaultCurrency = defaultCurrency;
        return this;
    }

    public HydraClientBuilder withLocale(String locale) {
        this.locale = locale;
        return this;
    }

    public HydraClientBuilder timeout(Duration timeout) {
        this.timeout = timeout;
        return this;
    }

    public HydraClientBuilder httpClient(HttpClient httpClient) {
        this.httpClient = httpClient;
        return this;
    }

    public HydraClient build() {
        return new HydraClientImpl(baseUrl, apiKey, secretKey, defaultCurrency, locale, timeout, httpClient);
    }
}

File: dotnet/src/HydraPayments.Sdk/HydraClientBuilder.cs
namespace HydraPayments.Sdk;

public class HydraClientBuilder
{
    private string _baseUrl = "http://localhost:8080";
    private string _apiKey = string.Empty;
    private string _secretKey = string.Empty;
    private string _defaultCurrency = "GBP";
    private string _locale = "en";
    private TimeSpan _timeout = TimeSpan.FromSeconds(30);
    private HttpClient? _httpClient;

    public HydraClientBuilder BaseUrl(string baseUrl)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        return this;
    }

    public HydraClientBuilder ApiKey(string apiKey)
    {
        _apiKey = apiKey;
        return this;
    }

    public HydraClientBuilder SecretKey(string secretKey)
    {
        _secretKey = secretKey;
        return this;
    }

    public HydraClientBuilder DefaultCurrency(string defaultCurrency)
    {
        _defaultCurrency = defaultCurrency;
        return this;
    }

    public HydraClientBuilder Locale(string locale)
    {
        _locale = locale;
        return this;
    }

    public HydraClientBuilder Timeout(TimeSpan timeout)
    {
        _timeout = timeout;
        return this;
    }

    public HydraClientBuilder HttpClient(HttpClient httpClient)
    {
        _httpClient = httpClient;
        return this;
    }

    public IHydraClient Build()
    {
        var httpClient = _httpClient ?? new HttpClient();
        httpClient.Timeout = _timeout;
        return new HydraClient(_baseUrl, _apiKey, _secretKey, _defaultCurrency, _locale, httpClient);
    }
}
